3221: update KEP according to 1.29 requirements (#4248)
kubernetes/enhancements: keps/sig-auth/3221-structured-authorization-configuration/README.md

kind: AuthorizationConfiguration
authorizers:
  - type: Webhook

>     name: system-crd-protector

Reviewer: huh, I missed that the design had the name at the authorizer entry level and the implementation had the name at the webhook config level.

I think we probably want a unique name to exist at the authorizer entry level as shown here, so that metrics / audit / logging can consistently use the name of the authorizer, regardless of the type.

We would probably make the name for built-in authz modes just be lower(mode), so `rbac`, `node`, `abac`, etc, and validate that in kube-apiserver specifically for the modes it knows about.

Reviewer 2: > We would probably make the name for built-in authz modes just be lower(mode), so rbac, node, abac, etc, and validate that in kube-apiserver specifically for the modes it knows about.

Wouldn't adding a new mode in the future be backwards incompatible then since it could conflict with the name of a webhook (and names need to be unique)?

Reviewer 2: Maybe not because you always have to opt-in to what authorizers you want to use?

Reviewer: yeah, not a problem, if you want to use a new built-in one, then you need to deconflict your names if we happen to have one that matches a name you chose for one of your webhooks

Reviewer: @palnabarun restore the name to the top-level type, and drop it from the webhook stanza, and we'll get a PR in place to make the code match

kind: AuthorizationConfiguration
authorizers:
  - name: system-webhook

>     name: system-crd-protector

Reviewer: nit: drop the double name

The user can define a CEL expression to determine whether a request needs to dispatched
to the authz webhook for which the expression has been defined. The user would have access
to a `request` variable containing a `SubjectAccessReview` object in the version specified
by the `matchConditionSubjectAccessReviewVersion` field. If the version specified by
`subjectAccessReviewVersion` in the request variable is `v1beta1`, the contents would be

> converted to the version specified in `matchConditionSubjectAccessReviewVersion` before

Reviewer 3: are we converting the subjectAccessReview object automatically to the version in matchConditionSubjectAccessReviewVersion or do we want to return a validation error if matchConditionSubjectAccessReviewVersion and subjectAccessReviewVersion don't match?

Reviewer: They're independent and don't have to match. You can send v1beta1 or v1 to the webhook. You can currently only set v1 for match conditions

for the Authorizer chain in both the old and new configuration scenarios.

We will add integration tests to validate the enablement/disablement flow.
  - When `--authorization-config` flags is defined, the feature flag must be turned on (when feature is in Alpha).
  - Setting `--authorization-config` along `--authorization-modes` and `--authorization-webhook-*`
command line flags should return an error.

>   - Configuring the authorizer using legacy flags should always be allowed

Reviewer: ```suggestion
  - Configuring the authorizer using legacy flags will continue to be allowed
```

Reviewer 3: do we need a test to ensure when cmdline flag values are updated between restarts, the corresponding values in the new structured authz config are updated?

Reviewer: Since the flag → structured config translation is internal and isn't persisted, as long as that has a unit test, that's probably sufficient
